### PR TITLE
Use hsa specialization for a fast path to memset.

### PR DESCRIPTION
**Pull request by iotamudelta**
This improves ROCm performance on an important customer benchmark by 20%.

Developed with @AlexVlx .

### CODE REVIEW

**Comment by iotamudelta** [comment]
@mangupta the test failures seem unrelated to this PR?

File: src/hip_memory.cpp [hunk 1]
```diff
@@ -1509,6 +1509,13 @@ __global__ void hip_copy2d_n(T* dst, const T* src, size_t width, size_t height,
 
 template <typename T>
 void ihipMemsetKernel(hipStream_t stream, T* ptr, T val, size_t sizeBytes) {
+
+    if(sizeof(T) == sizeof(uint32_t)){
+        // for all sizeof(uint32_t) use the fast path using hsa specialization
+        hsa_amd_memory_fill(ptr, reinterpret_cast<const std::uint32_t&>(val), sizeBytes);
```
Comment (mangupta) [inline src/hip_memory.cpp]: Failure is in one of the memset tests. I need to check the hsa spec, but I think the ```sizeBytes``` probably needs to be divided by ```sizeof(uint32_t)```.

Comment (iotamudelta) [reply src/hip_memory.cpp]: According to https://github.com/RadeonOpenCompute/ROCR-Runtime/blob/d823418d0500b30af775efb3b17b28e164958f3a/src/inc/hsa_ext_amd.h#L1282 you are right! Fixing.

Comment (iotamudelta) [reply src/hip_memory.cpp]: @mangupta looks like that's not sufficient to pass tests - mind having a look? Thanks!

**Comment by mangupta** [comment]
@iotamudelta The failures are still in all the memset related tests. Do all the HIP unit tests pass on your development system?
While I haven't tried digging into why the failures are occurring, there are few possible causes:
* Maybe you are hitting into either of the 2 failure cases @ https://github.com/RadeonOpenCompute/ROCR-Runtime/blob/master/src/inc/hsa_ext_amd.h#L1275-L1279.
* The call to hsa_amd_memory_fill ignores the stream i.e. the hsa_queue (which is important for the async memset cases). While it is doubtful that this can cause an issue, maybe it is worth checking.

**Pull request closed without merging**